### ISSUE TITLE
PEP 594: Remove stdlib modules scheduled for deletion in Python 3.13

### ISSUE THREAD

**Issue by vstinner** [python/cpython#104773]
[PEP 594](https://peps.python.org/pep-0594/) ("Removing dead batteries from the standard library") deprecated the following 19 modules in Python 3.11 (or before) and their removal in Python 3.13:

* [x] aifc
* [x] audioop
* [x] cgi
* [x] cgitb
* [x] chunk
* [x] crypt
* [x] imghdr
* [x] mailcap
* [x] msilib
* [x] nis
* [x] nntplib
* [x] ossaudiodev
* [x] pipes
* [x] sndhdr
* [x] spwd
* [x] sunau
* [x] telnetlib
* [x] uu
* [x] xdrlib

I propose to  remove them as soon as possible in the Python 3.13 development cycle to help users to prepare for this incompatible change before Python 3.13 final release (scheduled for October 2024).

I plan to work on pull requests to remove these modules.

### Linked PRs
* gh-104774
* gh-104775
* gh-104777
* gh-104778
* gh-104848
* gh-104862
* gh-104863
* gh-104864
* gh-104867
* gh-104868
* gh-104871
* gh-104894
* gh-104897
* gh-104900
* gh-104901
* gh-104908
* gh-104911
* gh-104927
* gh-104932
* gh-104933
* gh-104937
* gh-105139

**Comment by vstinner** [comment]
For bug triagers, I suppose that the next step is to close all issues related to removed modules.

**Comment by vstinner** [comment]
> Hi @vstinner, I can raise PRs if you allow, I will start with imghdr , someone referred me to this pep to avoid using imghdr package, I guess filetype is recommended for the same

I created the PR #104777 to remove imghdr, but there are issues with building the documentation (Sphinx). See the issue for details. The PR proposes alternatives in the What's New in Python 3.13 entry (``Doc/whatsnew/3.13.rst``).

**Comment by vstinner** [comment]
GitHub doesn't make it easy to have a chain of pull requests, so I prefer to merge my PRs as soon as possible to be able to work on the next module removal. One by one. The problem are conflicts on the ``Doc/whatsnew/3.13.rst`` file.

Status:

* PR #104777 removes imghdr but is blocked by Sphinx.
* PR #104868 removes chunk, but the aifc module should be removed first.
* test_aifc uses audioop, so aifc should be removed before audioop.
* I was scared by the [PEP 594: aifc section](https://peps.python.org/pep-0594/#aifc) which mentions users still relying on it. But since the PEP got approved by the SC, I think that I will be confident if a PR gets enough approvals from other core devs :-) Just, I chose to do it as late as possible :-)
* I plan to remove the uu module, but I was sad to see that we forgot to deprecated the 2 related functions in the binascii module, and so we will have to keep them for 2 more Python releases.

My plan (in which order I plan to remove remaining modules):

* PR #104908: Remove crypt
* Remove uu and deprecate the 2 related binascii functions
* Remove aifc
* Remove audioop

When the blocking issues are solved:

* PR #104777: Removes imghdr
* PR #104868: Removes chunk

@zware picked the removal of the msilib module (and _msi extension): PR #104911.

**Comment by vstinner** [comment]
> Remove aifc
> Remove audioop

By the way, I already have local branches removing them. I will rebase them when other modules will be removed.

**Comment by arhadthedev** [comment]
There are issues and PRs about the removed modules, however I don't know if we can close them until 3.12 stops accepting bugfixes.

cgi/cgib:
- gh-99747
- gh-104711

crypt:
- gh-77394
- gh-89966
- gh-98895

pipes:
- gh-82678

**Comment by vstinner** [comment]
So far, 18 modules have been removed. Removing the last one, **imghdr**, breaks the Docs CI job: PR #104777. I proposed to increase the minimum **Sphinx** version to build the Python documentation: issue #104818.

**Comment by vstinner** [comment]
Summary:

* All modules scheduled for removal by PEP 594 have been removed in Python 3.12 and Python 3.13
* The removal of the uu module makes 2 binascii functions useless and should be deprecated.
* One minor thing that we missed in PEP 594 was the removal of the uu codec: https://peps.python.org/pep-0594/#uu-and-the-uu-encoding
* I was wrong: we must not close all issues of removed modules, since bugs can still be fixed in Python 3.11 and 3.12: https://devguide.python.org/versions/

The remaining minor task related to the uu encoding is covered by PR #92758.

I'm not sure how to coordinate the work on helping third-party Python projects migrating to Python 3.13: move away from removed modules. But I prefer to not reuse this issue for this. Maybe discuss.python.org is a better place for that, since it's more visible than the Python bug tracker.

I created https://github.com/python/peps/pull/3156 to mark PEP 594 as Final.
